MapboxMobileEvents/MMEMetricsManager.m

```diff
@@ -0,0 +1,160 @@
+@property (nonatomic) int appWakeups;
+@property (nonatomic) int eventCountFailed;
+@property (nonatomic) int eventCountTotal;
+@property (nonatomic) int eventCountMax;
+@property (nonatomic) int deviceTimeDrift;
+@property (nonatomic) float deviceLat;
+@property (nonatomic) float deviceLon;
+@property (nonatomic) NSDate *dateUTC;
+@property (nonatomic) NSString *dateUTCString;
+@property (nonatomic) NSDictionary *configResponseDict;
+@property (nonatomic) NSMutableDictionary *eventCountPerType;
+@property (nonatomic) NSMutableDictionary *failedRequestsDict;
+
+@end
+
+@implementation MMEMetricsManager
+
++ (instancetype)sharedManager {
+    static MMEMetricsManager *_sharedManager;
+    static dispatch_once_t onceToken;
+    
+    dispatch_once(&onceToken, ^{
+        _sharedManager = [[MMEMetricsManager alloc] init];
+    });
+    
+    return _sharedManager;
+}
+
+- (void)metricsFromEventQueue:(NSArray *)eventQueue {
+    if (self.dateUTC == nil) {
+        [self updateDateUTC];
+    }
+    
+    if (eventQueue.count > 0) {
+        if (self.eventCountPerType == nil) {
+            self.eventCountPerType = [[NSMutableDictionary alloc] init];
+        }
+        
+        self.eventCountTotal = self.eventCountTotal + (int)eventQueue.count;
+        
+        for (MMEEvent *event in eventQueue) {
+            NSNumber *eventCount = [self.eventCountPerType objectForKey:event.name];
+            eventCount = [NSNumber numberWithInteger:[eventCount integerValue] + 1];
+            [self.eventCountPerType setObject:eventCount forKey:event.name];
+        }
+    }
+}
+
+- (void)metricsFromEvents:(nullable NSArray *)events error:(nullable NSError *)error {
+    if (error == nil) {
+        self.requests = self.requests + 1;
+    } else {
+        if (events) {
+            self.eventCountFailed = self.eventCountFailed + (int)events.count;
+        }
+        
+        if ([error.userInfo objectForKey:MMEResponseKey]) {
+            NSHTTPURLResponse *response = (NSHTTPURLResponse *)[error.userInfo objectForKey:MMEResponseKey];
+            NSString *urlString = [[response URL] absoluteString];
+            NSNumber *statusCode = [NSNumber numberWithLong:(long)response.statusCode];
+            NSString *statusCodeString = [statusCode stringValue];
+            NSString *failedRequestKey = [NSString stringWithFormat:@"%@, %@",urlString, statusCodeString];
+            
+            if (self.failedRequestsDict == nil) {
+                self.failedRequestsDict = [[NSMutableDictionary alloc] init];
+            }
+        
+            NSNumber *failedCount = [self.failedRequestsDict objectForKey:failedRequestKey];
+            failedCount = [NSNumber numberWithInteger:[failedCount integerValue] + 1];
+            [self.failedRequestsDict setObject:failedCount forKey:failedRequestKey];
+        }
+    }
+}
+
+- (void)metricsFromData:(NSData *)data {
+    self.totalDataTransfer = self.totalDataTransfer + data.length;
```

Comment:
```suggestion
    self.totalDataTransfer += data.length;
```
would this be cleaner? same question on line 90, 92, 97

Reply:
Yes!